File: gestion-projet-frontend/src/app/core/services/meeting.service.ts
import { HttpClient } from '@angular/common/http';
import { Injectable } from '@angular/core';
import { Observable } from 'rxjs';
import {
  AttendeeResponse,
  Meeting,
  MeetingRequest,
  MeetingStatus
} from 'src/app/shared/models/meeting.model';

@Injectable({
  providedIn: 'root'
})
export class MeetingService {
  private readonly api = '/api/meetings';

  constructor(private http: HttpClient) {}

  create(req: MeetingRequest): Observable<Meeting> {
    return this.http.post<Meeting>(this.api, req);
  }

  update(id: number, req: MeetingRequest): Observable<Meeting> {
    return this.http.put<Meeting>(`${this.api}/${id}`, req);
  }

  delete(id: number): Observable<void> {
    return this.http.delete<void>(`${this.api}/${id}`);
  }

  getById(id: number): Observable<Meeting> {
    return this.http.get<Meeting>(`${this.api}/${id}`);
  }

  getByProject(projectId: number): Observable<Meeting[]> {
    return this.http.get<Meeting[]>(`${this.api}/project/${projectId}`);
  }

  /** Reunions a venir ou je suis invite ou organisateur. */
  getMyMeetings(): Observable<Meeting[]> {
    return this.http.get<Meeting[]>(`${this.api}/me`);
  }

  /** Toutes mes reunions (passees + futures). */
  getAllMyMeetings(): Observable<Meeting[]> {
    return this.http.get<Meeting[]>(`${this.api}/me/all`);
  }

  changeStatus(id: number, status: MeetingStatus): Observable<Meeting> {
    return this.http.patch<Meeting>(`${this.api}/${id}/status`, { status });
  }

  updateNotes(id: number, notes: string): Observable<Meeting> {
    return this.http.patch<Meeting>(`${this.api}/${id}/notes`, { notes });
  }

  respond(id: number, response: AttendeeResponse): Observable<Meeting> {
    return this.http.post<Meeting>(`${this.api}/${id}/respond`, { response });
  }

  /** Rejoint la reunion en ligne : passe IN_PROGRESS et retourne roomName. */
  joinMeeting(id: number): Observable<Meeting> {
    return this.http.post<Meeting>(`${this.api}/${id}/join`, {});
  }
}

File: GestionProjet/src/main/java/com/example/gestionprojet/services/interfaces/MeetingService.java
package com.example.gestionprojet.services.interfaces;

import com.example.gestionprojet.dto.MeetingRequest;
import com.example.gestionprojet.dto.MeetingResponse;
import com.example.gestionprojet.entities.AttendeeResponse;
import com.example.gestionprojet.entities.MeetingStatus;

import java.util.List;

public interface MeetingService {

    MeetingResponse create(MeetingRequest request);

    MeetingResponse update(Long id, MeetingRequest request);

    void delete(Long id);

    MeetingResponse getById(Long id);

    List<MeetingResponse> getByProject(Long projectId);

    List<MeetingResponse> getMyUpcomingMeetings();

    /** Toutes mes reunions (passees + futures), sans filtre de date. */
    List<MeetingResponse> getAllMyMeetings();

    MeetingResponse changeStatus(Long id, MeetingStatus status);

    MeetingResponse updateNotes(Long id, String notes);

    /** Pour un invite : repond a l'invitation. */
    MeetingResponse respond(Long meetingId, AttendeeResponse response);

    /**
     * Rejoint la reunion en ligne : passe le statut a IN_PROGRESS si SCHEDULED
     * et retourne la reponse enrichie (avec roomName).
     * Accessible a l'organisateur et aux invites ayant accepte.
     */
    MeetingResponse joinMeeting(Long id);
}
